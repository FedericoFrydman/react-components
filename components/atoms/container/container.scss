@import '../../assets/styles/colors.scss';

.lsux-container {

  &.lsux-container--light-gray {
    background-color: $N100;
  }
  &.lsux-container--white {
    background-color: $N00;
  }
  &.lsux-container--fixed {
    position: fixed;
    z-index: 99;
  }
  &.lsux-container--flexbox {
    display: flex;

    &.lsux-container--flex-direction-row {
      flex-direction: row;
    }
    &.lsux-container--flex-direction-row-reverse {
      flex-direction: row-reverse;
    }
    &.lsux-container--flex-direction-column {
      flex-direction: column;
    }
    &.lsux-container--flex-direction-column-reverse {
      flex-direction: column-reverse;
    }

    &.lsux-container--flex-wrap-wrap {
      flex-wrap: wrap;
    }
    &.lsux-container--flex-wrap-nowrap {
      flex-wrap: nowrap;
    }
    &.lsux-container--flex-wrap-reverse {
      flex-wrap: wrap-reverse;
    }

    &.lsux-container--flex-justify-flex-start {
      justify-content: flex-start;
    }
    &.lsux-container--flex-justify-flex-end {
      justify-content: flex-end;
    }
    &.lsux-container--flex-justify-center {
      justify-content: center;
    }
    &.lsux-container--flex-justify-space-between {
      justify-content: space-between;
    }
    &.lsux-container--flex-justify-space-around {
      justify-content: space-around;
    }

    &.lsux-container--flex-items-flex-start {
      align-items: flex-start;
    }
    &.lsux-container--flex-items-flex-end {
      align-items: flex-end;
    }
    &.lsux-container--flex-items-center {
      align-items: center;
    }
    &.lsux-container--flex-items-baseline {
      align-items: baseline;
    }
    &.lsux-container--flex-items-stretch {
      align-items: stretch;
    }

    &.lsux-container--flex-content-flex-start {
      align-content: flex-start;
    }
    &.lsux-container--flex-content-flex-end {
      align-content: flex-end;
    }
    &.lsux-container--flex-content-center {
      align-content: center;
    }
    &.lsux-container--flex-content-space-between {
      align-content: space-between;
    }
    &.lsux-container--flex-content-space-around {
      align-content: space-around;
    }
    &.lsux-container--flex-content-strech {
      align-content: stretch;
    }
  }
}
