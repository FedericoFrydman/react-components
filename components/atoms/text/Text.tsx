import { forwardRef, ReactElement, Ref } from 'react';
import cn from 'classnames';

import './text.scss';
import '../../assets/styles/utilities.scss';

export type TextAs = 'p' | 'span';

export type TextSize = 'body' | 'description' | 'tiny' | 'small' | 'medium' | 'large' | 'extra-large';

export type TextWeight = 'normal' | 'semibold' | 'bold';

export type TextColor = 'N1000' | 'N800' | 'N700' | 'N500' | 'N300' | 'N100' | 'N00';

export interface TextProps extends Omit<React.HTMLProps<HTMLParagraphElement>, 'css'> {
  /**
   * Render a span or a paragraph. Default 'span'.
   */
  as?: TextAs;
  /**
   * Text value to be rendered
   */
  text?: string;
  /**
   * Defines the text height
   */
  textHeight?: string;
  /**
   * Size of the text. Defaults to `medium`.
   */
  textSize?: TextSize;
  /**
   * Weight of the text
   */
  textWeight?: TextWeight;
  /**
   * Sets the `textWeight` to bold
   * @deprecated Use `fontWeight='bold'` instead. Will be removed in future verison.
   */
  bold?: boolean;
  /**
   * Sets the `textWeight` to semibold
   * @deprecated Use `fontWeight='semibold'` instead. Will be removed in future verison.
   */
  semiBold?: boolean;
  /**
   * Sets the text style to italic.
   */
  italic?: boolean;
  /**
   * Add a strikethrough to the text.
   */
  strike?: boolean;
  /**
   * Text value to be rendered
   */
  children?: string | number | ReactElement;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Color of the text
   * @deprecated Text defaults to N800 color. This prop will be removed in a future verison.
   */
  color?: TextColor;
  /**
   * Modifies the text as separator
   */
  separator?: boolean;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLParagraphElement>;
}

const TextComponent = (
  {
    as,
    bold,
    children,
    color,
    italic,
    semiBold,
    separator,
    strike,
    text,
    textHeight,
    textSize = 'medium',
    textWeight = 'normal',
    classNames = [],
    ...props
  }: TextProps,
  ref: TextProps['ref'],
) => {
  // Warn if bold or semibold props are used
  if (bold || semiBold) {
    console.warn('bold & semibold are deprecated. Please use `textWeight` instead.');
  }

  // Warn if color prop is used
  if (color) {
    console.warn('color is deprecated. Text is defaulted to $N800');
  }

  // Warn if body or description textSize values are used
  if (textSize === 'body' || textSize === 'description') {
    console.warn(
      'textSizes `body` and `description` have been deprecated. Please use `tiny`, `small`, `medium`, `large`, or `extra-large` instead.',
    );
  }

  // Set textWeight
  if (semiBold) {
    textWeight = 'semibold';
  }
  if (bold) {
    textWeight = 'bold';
  }

  const classText = 'lsux-text';
  const setHeight = { height: textHeight, overflow: 'auto' };
  const useHeight = textHeight ? setHeight : {};

  const cnText = cn(
    {
      [`${classText}`]: true,
      [`${classText}--separator`]: separator,
      [`${classText}--italic`]: italic,
      [`${classText}--strike`]: strike,
      [`${classText}--${textSize}`]: textSize,
      [`${classText}--${textWeight}`]: textWeight,
    },
    classNames,
  );

  if (as === 'span') {
    return (
      <span {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </span>
    );
  } else {
    return (
      <p {...props} ref={ref} style={{ ...useHeight, ...props.style }} className={cnText}>
        {text ? text : children}
      </p>
    );
  }
};

export const Text = forwardRef<HTMLParagraphElement, TextProps>(TextComponent);
