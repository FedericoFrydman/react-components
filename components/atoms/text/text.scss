@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/colors.scss';

.lsux-text {
  margin: 0;

  // Text Weights
  &--normal {
    font-weight: normal;
  }
  &--bold {
    font-weight: bold;
  }
  &--semibold {
    font-weight: 600;
  }

  // Text Sizes
  &--tiny {
    font-size: 10px;
    font-style: normal;
    letter-spacing: -0.01em;
    line-height: 12px;
  }
  &--description,
  &--small {
    font-size: 12px;
    font-style: normal;
    letter-spacing: -0.006em;
    line-height: 16px;
  }
  &--body,
  &--medium {
    font-size: 14px;
    font-style: normal;
    letter-spacing: -0.006em;
    line-height: 18px;
  }
  &--large {
    font-size: 16px;
    font-style: normal;
    letter-spacing: -0.011em;
    line-height: 22px;
  }
  &--extra-large {
    font-size: 18px;
    font-style: normal;
    letter-spacing: -0.014em;
    line-height: 24px;
  }

  // Text Styles
  &--italic {
    font-style: italic;
  }
  &--strike {
    text-decoration: line-through;
  }

  &--no-margin {
    margin: 0;
  }

  &--separator {
    align-items: center;
    display: flex;
    text-transform: lowercase;

    &::before,
    &::after {
      background-color: $N300;
      content: '';
      flex-grow: 1;
      height: 1px;
    }

    &::before {
      margin-right: 16px;
    }

    &::after {
      margin-left: 16px;
    }
  }
}
