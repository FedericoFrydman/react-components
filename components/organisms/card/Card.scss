@import '../../assets/styles/colors.scss';
@import '../../assets/styles/fonts.scss';
@import '../../assets/styles/shadows.scss';

.lsux-card {
  border-radius: 11px;
  box-shadow: 0px -1px 12px rgba(0, 0, 0, 0.015), 0px 6px 8px rgba(0, 0, 0, 0.08);
  display: flex;
  flex-direction: column;
  transition: all 0.15s ease-in-out;

  &--bg {
    &--light-gray {
      background: $N200;
    }
    &--white {
      background: $N00;
    }
    &--none {
      background: transparent;
    }
  }

  &--selectable:not(:disabled) {
    cursor: pointer;
  }

  &--selected {
    box-shadow: 0px -1px 12px rgba(0, 0, 0, 0.015), 0px 6px 8px rgba(0, 0, 0, 0.08),
      0 0 0 1.5px var(--brand-color-prim300);
    .lsux-card__image > img {
      border-top-left-radius: 8px;
      border-top-right-radius: 8px;
    }
  }

  &--unselected {
    .lsux-card__image > img {
      border-top-left-radius: 11px;
      border-top-right-radius: 11px;
    }
  }

  &__content {
    background: none;
    border-radius: 11px;
  }

  &__image > img {
    border-top-left-radius: 8px;
    border-top-right-radius: 8px;
    height: 100%;
    object-fit: cover;
    width: 100%;
  }
  &.lsux-card--bg--none .lsux-card__image > img {
    border-radius: 11px;
  }

  &--card-actions {
    display: flex;
    flex-direction: row;
    margin: 0.75rem;
    .lsux-button {
      padding: 0;
      &:not(:last-child) {
        margin-right: 0.3rem;
      }
    }
  }
}
