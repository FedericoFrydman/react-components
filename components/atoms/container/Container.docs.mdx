import { ArgsTable, Story, Canvas, Stories } from '@storybook/addon-docs';
import { Alert } from '../../molecules/alert/Alert';
import { Container } from './Container';

# Container

<Canvas>
  <Story id="layout-container--light-gray-background" />
</Canvas>

<Alert title="Dev Note" dismissible={false} severity="warning" icon="alert_warning">
  This component is just to use it when the child container has special specifications, like background, position,
  etc...
</Alert>

<p style={{ margin: 0 }}>&nbsp;</p>

<ArgsTable of={Container}/>

## Example Code

```tsx
<Container background="light-gray" fixed={true}>
  // the children go here
<Container/>
```

## Usage Guideline

| Prop       | Usage                                                                                                                                                             |
| ---------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| background | This prop should be equal to "light-gray" in order to add a class to the container that will provied the background color. Now it only support light-gray ($N100) |
| fixed      | When it is true, it adds a class to the container to make it position fixed with 99 z-index                                                                       |
| flexbox    | When it is true, it adds a class to the container to make it display flex                                                                                         |
