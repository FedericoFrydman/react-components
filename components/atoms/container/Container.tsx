import { forwardRef, Ref, ReactElement } from 'react';
import cn from 'classnames';

import './container.scss';

export type ContainerBgColors = 'light-gray' | 'white' | 'none';

export interface ContainerProps extends Omit<React.HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Define if container has background
   * @default 'white'
   */
  background?: ContainerBgColors;
  /**
   * Define if container has position fixed
   */
  fixed?: boolean;
  /**
   * Define if container has display flex
   */
  flexbox?: boolean;
  /**
   * Define the flex-direction of flex-container
   */
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse';
  /**
   * Define the flex-wrap of flex-container
   */
  flexWrap?: 'wrap' | 'nowrap' | 'wrap-reverse';
  /**
   * Define the flex-direction
   */
  justifyContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around';
  /**
   * Define the align-items for flex-container
   */
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch';
  /**
   * Define the flex-direction
   */
  alignContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around' | 'stretch';
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Children Container
   */
  children?: string | number | ReactElement | ReactElement[];
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const ContainerComponent = (
  {
    background = 'white',
    classNames = [],
    children,
    fixed,
    flexbox,
    flexDirection,
    flexWrap,
    justifyContent,
    alignItems,
    alignContent,
    ...props
  }: ContainerProps,
  ref: ContainerProps['ref'],
) => {
  const classContainer = 'lsux-container';
  const cnContainer = cn(
    {
      [`${classContainer}`]: true,
      [`${classContainer}--${background}`]: background !== 'none',
      [`${classContainer}--fixed`]: fixed,
      [`${classContainer}--flexbox`]: flexbox,
      [`${classContainer}--flex-direction-${flexDirection}`]: flexDirection,
      [`${classContainer}--flex-wrap-${flexWrap}`]: flexWrap,
      [`${classContainer}--flex-justify-${justifyContent}`]: justifyContent,
      [`${classContainer}--flex-items-${alignItems}`]: alignItems,
      [`${classContainer}--flex-content-${alignContent}`]: alignContent,
    },
    classNames,
  );

  return (
    <div {...props} className={cnContainer} ref={ref}>
      {children}
    </div>
  );
};

export const Container = forwardRef<HTMLDivElement, ContainerProps>(ContainerComponent);
