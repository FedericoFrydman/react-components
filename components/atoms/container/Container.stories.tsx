import { Story, Meta } from '@storybook/react';
import '../../assets/styles/utilities.scss';

import docs from './Container.docs.mdx';
import { Container, ContainerProps } from './Container';
import { Text } from '../text/Text';

const ContainerStory: Story<ContainerProps> = (args: ContainerProps) => (
  <Container {...args}>
    <Text text="I am container child!" as="p" classNames={['p-5']} />
  </Container>
);

export const WhiteBackground = ContainerStory.bind({});
WhiteBackground.args = {
  background: 'white',
};

export const LightGrayBackground = ContainerStory.bind({});
LightGrayBackground.args = {
  background: 'light-gray',
};

export default {
  argTypes: {
    background: {
      options: ['light-gray', 'white', 'none'],
      type: 'select',
    },
  },
  component: ContainerStory,
  parameters: {
    docs: { page: docs },
  },
  title: 'Layout/Container',
} as Meta;
