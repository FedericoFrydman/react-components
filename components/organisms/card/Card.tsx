import { Children, cloneElement, FC, forwardRef, isValidElement, ReactElement, ReactNode, Ref } from 'react';
import cn from 'classnames';

import { Button, ButtonProps, ButtonVariant } from '../../molecules/button/Button';
import { Container, ContainerBgColors, ContainerProps } from '../../../components/atoms/container/Container';
import { Image } from '../../../components/atoms/image/Image';
import { Heading } from '../../../components/atoms/heading/Heading';
import { Text } from '../../../components/atoms/text/Text';
import '../../assets/styles/utilities.scss';
import './Card.scss';

const classCard = 'lsux-card';

export type PaddingType = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

export interface DeprecatedCardProps {
  /**
   * Title of the card
   * @deprecated Please use `Card.Title` instead.
   */
  title?: string;
  /**
   * Define the src of the optional image
   * @deprecated Please use `Card.Image` instead.
   */
  imageURL?: string;
  /**
   * Text content in the card
   * @deprecated Please use `Text` instead.
   */
  text?: string;
  /**
   * Text in the (optional) button
   * @deprecated Please use `Card.Actions` instead.
   */
  buttonText?: string;
  /**
   * Variant of the button
   * @deprecated Please use `Card.Actions` instead.
   */
  buttonVariant?: ButtonVariant;
  /**
   * Custom padding
   * @deprecated Please use `classNames` in `Card.Content` instead.
   */
  padding?: PaddingType;
  /**
   * Define card title height
   * @deprecated This will be removed in a future version.
   */
  cardTitleHeight?: string;
  /**
   * Define card text height
   * @deprecated No longer works. Use `Text` component instead. This will be removed in a future version.
   */
  cardTextHeight?: string;
  /**
   * Alt text for image
   * @deprecated Please use `Card.Image` instead.
   */
  imageAltText?: string;
}

export interface CardProps extends DeprecatedCardProps, Omit<React.HTMLProps<HTMLDivElement>, 'onClick' | 'css'> {
  children?: ReactNode | ReactNode[];
  /**
   * Define card height
   */
  cardHeight?: string;
  /**
   * Defines card width.
   */
  cardWidth?: string;
  /**
   * Define if container has background
   * @default white
   */
  background?: ContainerBgColors;
  /**
   * Determines if the card can be selected or not. This adds selectable styling and pointer.
   */
  selectable?: boolean;
  /**
   * Determines if the card is selected or not
   */
  isSelected?: boolean;
  /**
   * Additional class names
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
  /**
   * Callback function when card is clicked
   */
  onClick?: (id: string) => void;
}

export interface CardActionsProps {
  /**
   * Alignment of the buttons
   * @default left
   */
  align?: 'left' | 'center' | 'right';
  /**
   * Buttons to be displayed
   */
  children: ReactElement | ReactElement[];
}
export const CardActions: FC<CardActionsProps> = ({ children, align = 'left' }: CardActionsProps) => {
  const childrenWithProps = Children.map(children as ReactElement<ButtonProps>[], (child) => {
    if (isValidElement(child)) {
      return cloneElement(child, { shape: 'round', textSize: 'medium', variant: 'tertiary' });
    }
    return child;
  });

  const alignMap: Record<string, string> = {
    center: 'justify-content-center',
    left: 'justify-content-start',
    right: 'justify-content-end',
  };

  const cnCardAction = cn(`${classCard}--card-actions`, alignMap[align]);

  return <div className={cnCardAction}>{childrenWithProps}</div>;
};

export interface CardContentProps extends ContainerProps {
  /**
   * Define if container has background
   */
  background?: ContainerBgColors;
  /**
   * Elements to show in the card content
   */
  children: ReactElement | ReactElement[];
  /**
   * Additional class names
   */
  classNames?: cn.Argument;
  /**
   * Padding for the content
   * @deprecated Please use `classNames` instead
   */
  padding?: number;
}
export const CardContent: FC<CardContentProps> = ({
  children,
  classNames = [],
  padding = 4,
  background,
  ...props
}: CardContentProps) => {
  const cnCardContent: cn.Argument = [`${classCard}__content`, `p-${padding}`, classNames];
  return (
    <Container {...props} background={background} classNames={cnCardContent}>
      {children}
    </Container>
  );
};

export interface CardImageProps {
  imageURL: string;
  imageAltText?: string;
  classNames?: cn.Argument;
}
export const CardImage: FC<CardImageProps> = ({ imageURL, imageAltText, classNames = [] }: CardImageProps) => {
  const cnCardImage = cn(`${classCard}__image`, classNames);
  return (
    <div className={cnCardImage}>
      <Image src={imageURL} alt={imageAltText} fluid />
    </div>
  );
};

export interface CardTitleProps {
  /**
   * Title of the card
   */
  children: string;
  /**
   * Additional class names
   */
  classNames?: cn.Argument;
  /**
   * Card title height
   */
  titleHeight?: string;
}
export const CardTitle: FC<CardTitleProps> = ({ children, classNames = [], titleHeight }: CardTitleProps) => {
  const cnCardTitle = cn(`${classCard}__title`, classNames);
  return (
    <div className={cnCardTitle}>
      <Heading as="T16" text={children} titleHeight={titleHeight} classNames={['pb-3']} />
    </div>
  );
};

/**
 * Cards are flexible, rectangular containers for text, images, or actions for a single topic. Cards are intended to be flexible across device widths and should live within our grid and defined layout.
 */
const CardComponent = forwardRef<HTMLDivElement, CardProps>(
  (
    {
      onClick = null,
      children,
      classNames = [],
      buttonText,
      buttonVariant,
      cardHeight,
      cardTextHeight,
      cardTitleHeight,
      cardWidth,
      background,
      imageAltText,
      imageURL,
      padding,
      isSelected,
      selectable = isSelected !== undefined ?? false,
      text,
      title,
      ...props
    },
    ref: CardProps['ref'],
  ) => {
    // Warn for deprecated props
    if (title) console.warn(`title has been deprecated.`);
    if (imageURL) console.warn(`imageURL has been deprecated.`);
    if (text) console.warn(`text has been deprecated.`);
    if (buttonText) console.warn(`buttonText has been deprecated.`);
    if (buttonVariant) console.warn(`buttonVariant has been deprecated.`);
    if (padding) console.warn(`padding has been deprecated.`);
    if (cardTitleHeight) console.warn(`cardTitleHeight has been deprecated.`);
    if (cardTextHeight) console.warn(`cardTextHeight has been deprecated.`);
    if (imageAltText) console.warn(`imageAltText has been deprecated.`);

    // Build content when deprecated props are used
    let cardContent = null;
    let cardImage = null;
    let cardTitle = null;
    let cardText = null;

    if (imageURL) cardImage = <CardImage imageURL={imageURL} imageAltText={imageAltText} />;
    if (title) cardTitle = <CardTitle titleHeight={cardTitleHeight}>{title}</CardTitle>;
    if (text) cardText = <Text text={text} />;

    // Backwards compatibility for cardTitle & cardText
    if (cardTitle || cardText) {
      cardContent = (
        <CardContent classNames={[`p-${padding}`]}>
          {cardTitle}
          {cardText}
        </CardContent>
      );
    }

    // Backwards compatibility card actions
    let cardActions = null;
    if (buttonText)
      cardActions = (
        <Card.Actions>
          <Button shape="round" label={buttonText} variant={buttonVariant} />
        </Card.Actions>
      );

    const handleCardClick = (): void => {
      if (onClick != null) {
        onClick(props.id);
      }
    };

    const cnCard = cn(
      {
        [`${classCard}`]: true,
        [`${classCard}--bg--${background ?? 'white'}`]: true,
        [`${classCard}--selectable`]: selectable,
        [`${classCard}--selected`]: isSelected,
        [`${classCard}--unselected`]: !isSelected,
      },
      classNames,
    );

    const cardStyle = {
      height: cardHeight,
      width: cardWidth,
    };

    return (
      <div {...props} ref={ref} onClick={handleCardClick} className={cnCard} style={{ ...cardStyle, ...props.style }}>
        {cardImage}
        {cardContent}
        {cardActions}
        {children}
      </div>
    );
  },
) as FC<CardProps>;

CardComponent.displayName = 'Card';

type CardActionsComponent = FC<CardActionsProps>;
type CardContentComponent = FC<CardContentProps>;
type CardImageComponent = FC<CardImageProps>;
type CardTitleComponent = FC<CardTitleProps>;
type CardComp = FC<CardProps> & {
  Actions: CardActionsComponent;
  Content: CardContentComponent;
  Image: CardImageComponent;
  Title: CardTitleComponent;
};

export const Card = CardComponent as CardComp;

Card.Actions = CardActions;
Card.Content = CardContent;
Card.Image = CardImage;
Card.Title = CardTitle;
