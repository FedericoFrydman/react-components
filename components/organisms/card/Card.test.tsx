import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import { Card, CardProps } from './Card';
import { Text } from '../../atoms/text/Text';
import { ContainerBgColors } from '../../../components/atoms/container/Container';

describe('Card Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Card>
        <Card.Content>
          <Text text="Example text" />
        </Card.Content>
      </Card>,
    );
    testId = 'test card';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the card the provided classNames', () => {
          const card = renderCard({ classNames });

          expect(card).toHaveClass(classNames.join(' '), { exact: false });
        });
      });
    });

    describe('selected', () => {
      const selectedClassName = ['lsux-card--selectable', 'lsux-card--selected'].join(' ');

      it('gives the card the selected class', () => {
        const card = renderCard({ isSelected: true });

        expect(card).toHaveClass(selectedClassName, { exact: false });
      });
    });

    describe('title', () => {
      const title = 'my card title';

      it('is displayed when present', () => {
        renderCard({ title });

        expect(screen.getByText(title)).toBeVisible();
      });
    });

    describe('background', () => {
      const backgrounds: ContainerBgColors[] = ['light-gray', 'white', 'none'];

      it('defaults to white', () => {
        const card = renderCard();

        expect(card).toHaveClass(`lsux-card--bg--white`);
      });

      function itAssignsTheRightClassForBackground(background: ContainerBgColors) {
        const cardBgClass = `lsux-card--bg--${background}`;

        describe(`when the background is ${background}`, () => {
          it(`assigns the ${cardBgClass} to the card`, () => {
            const card = renderCard({ background: background });

            expect(card).toHaveClass(cardBgClass);
          });
        });
      }

      backgrounds.forEach((bg) => itAssignsTheRightClassForBackground(bg));
    });

    describe('cardHeight', () => {
      describe('when "cardHeight" is provided', () => {
        const cardHeight = `200px`;

        it('sets the card width to the provided value', () => {
          const card = renderCard({ cardHeight });

          expect(card).toHaveStyle({ height: cardHeight });
        });
      });
    });

    describe('cardWidth', () => {
      describe('when "cardWidth" is provided', () => {
        const cardWidth = `200px`;

        it('sets the card width to the provided value', () => {
          const card = renderCard({ cardWidth });

          expect(card).toHaveStyle({ width: cardWidth });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the card', () => {
      expect(renderCard({ ref })).toEqual(ref.current);
    });
  });

  describe('unspecified html attributes', () => {
    const altText = 'My accessible card';

    it('allows html attributes to be passed to the underlying card element', () => {
      render(<Card alt={altText} data-testid={testId} />);

      const card = screen.getByTestId(testId);

      expect(card).toHaveAttribute('alt', altText);
    });
  });

  const renderCard = (props: CardProps = {}): HTMLElement => {
    render(<Card {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
