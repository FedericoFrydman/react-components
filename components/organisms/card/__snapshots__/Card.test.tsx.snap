// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component should render the component 1`] = `
<div
  className="lsux-card lsux-card--bg--white lsux-card--unselected"
  onClick={[Function]}
  style={
    Object {
      "height": undefined,
      "width": undefined,
    }
  }
>
  <CardContent>
    <ForwardRef(TextComponent)
      text="Example text"
    />
  </CardContent>
</div>
`;
